### DISCUSSION TITLE
GTFS editing capabilities

### DISCUSSION THREAD

**Discussion by markusstraub** [araichev/gtfs_kit#35] [General]
Hi,

building on your nifty library I implemented some GTFS editing capabilities in https://github.com/ait-energy/gtfs-fiddler/. Quoting the README:

---

1. Add additional trips
   - Earliest trip in the morning (for a specific time) with `GtfsFiddler.ensure_earliest_departure`
   - Latest trip in the evening (for a specific time) with `GtfsFiddler.ensure_latest_departure`
   - Trips to shorten intervals (for a specified maximum interval duration) with `GtfsFiddler.ensure_max_trip_interval`
2. Increase speed of trips (for a specified average speed between two stops) with `GtfsFiddler.ensure_min_speed`

Also it provides typed access to the more of the feed's members (for autocompletion in IDE :)

The helper method `fiddle.compute_stop_time_stats` supplements the gtfs_kit utils.

---

Are you interested in merging this into gtfs_kit? I think it would be a great addition and I am happy to provide the code. Tell me what you think!

Best, Markus

**Comment by araichev** [comment]
Hi Markus, good to hear you're finding GTFS Kit useful.
Skimming over your GTFS Fiddler code, i see how it can also be useful.
However, i also think it's a little too niche to add to GTFS Kit, which is meant to be a general purpose GTFS library.
At this point, i think Fiddler is better maintained as a separate library.

**Comment by markusstraub** [comment]
Hi Alex, OK let's keep it separate. If you change your mind I'm happy to help with the integration.

However, I would still suggest integration of [compute_stop_time_stats](https://github.com/ait-energy/gtfs-fiddler/blob/c37e51274fd1795ad80495521ecf6abcb5c909f9/src/gtfs_fiddler/fiddle.py#L33) into gtfs_kit as this is a general-purpose analysis method, what do you think?

**Comment by diegoperezalvarez** [comment]
Hello,

Related point, what would you think of gtfs merging capabilities? I.e. merge all subfiles on *_id columns, possibly with a few optional cleaning steps before and after the merge (like extend_id, say, or possibly an option to exclude optional gtfs files). Happy to write up a PR if this is useful.

**Comment by araichev** [comment]
> Hello,
> 
> Related point, what would you think of gtfs merging capabilities? I.e. merge all subfiles on *_id columns, possibly with a few optional cleaning steps before and after the merge (like extend_id, say, or possibly an option to exclude optional gtfs files). Happy to write up a PR if this is useful.

Definitely maybe.
I don't ever use merging myself.
What's the use case, @diegoperezalvarez ?

**Comment by diegoperezalvarez** [comment]
At my job, I am often receiving disparate GTFS feeds from different places.
I sometimes am tasked with joining these feeds together (for example, a bus and rail feed).
After merging, there are typically a few conflicts which may need resolving. To give an example of a recent conflict I had:

1. I restricted a national rail GTFS file to a local context with `restrict_to_area`.
2. I had merged the resultant rail + bus feed (by accessing and merging each subfile), prepending each `_id` field with "R" and "B" to ensure no conflicts when merged.

Conflicts:

- Certain stops had not parent station in the feed (rail issue, more so with `restrict_to_area` but an easy fix).
- Some optional gtfs files had lost reference (due to missing stop and missing _id due to the prepend).

**Comment by araichev** [comment]
OK @diegoperezalvarez , it sounds like the feature could be useful to users of GTFS Kit.
If you're still up for the task, please go ahead with a merge request into develop that implements the feature and includes tests.
I suggest reading up on [Google's old GTFS merge tool](https://github.com/google/transitfeed/wiki/Merge) before you start your design.

**Comment by diegoperezalvarez** [comment]
Thanks @araichev - will need to wait for another job to come round with a similar requirement so I can justify the time to put into it. Putting a link to the source [here](https://github.com/google/transitfeed/blob/master/merge.py) to read through later.
